README.md
# Flutter Bouncing Text

A Flutter package to create a Romantic Bouncing Text. It will be useful for your awesome app.

More packages by [Romantic Developer](https://pub.dev/publishers/romanticdeveloper.com/packages)

![Demo](./demo.gif)

### Example

```dart
import 'package:flutter/material.dart';
import 'package:flutter_bouncing_text/flutter_bouncing_text.dart';

void main() => runApp(MyApp());

class MyApp extends StatefulWidget {
  @override
  _MyAppState createState() => _MyAppState();
}

class _MyAppState extends State<MyApp> {
  @override
  Widget build(BuildContext context) {
    return MaterialApp(
      title: 'Example',
      theme: ThemeData(
        primarySwatch: Colors.blue,
      ),
      home: Home(),
    );
  }
}

class Home extends StatefulWidget {
  const Home({Key? key}) : super(key: key);

  @override
  State<Home> createState() => _HomeState();
}

class _HomeState extends State<Home> {
  
  @override
  Widget build(BuildContext context) {
    return Scaffold(
      appBar: AppBar(
        title: Text('Romantic Developer'),
      ),
      body: SafeArea(
        child: SingleChildScrollView(
          padding: const EdgeInsets.all(20.0),
          child: Column(
            crossAxisAlignment: CrossAxisAlignment.stretch,
            children: [
                AnimatedBouncingText(
                  text: 'I am\nRomantic Developer.',
                  mode: BouncingTextModes.sequenceOneTime,
                  textAlign: TextAlign.center,
                  characterDuration: const Duration(seconds: 1),
                  characterDelay: const Duration(milliseconds: 300),
                  textStyle: TextStyle(
                    color: Colors.white,
                    fontWeight: FontWeight.bold,
                    fontSize: 22,
                  ),
                  onEnd: () {
                    print('_HomeState.build.onEnd.$index');
                  },
                ),
            ],
          ),
        ),
      ),
    );
  }
}

```

### Development environment

```
[✓] Flutter (Channel stable, 3.0.5, on macOS 12.6 21G115 darwin-x64, locale en-VN)
    • Flutter version 3.0.5 at ~/fvm/versions/3.0.5
    • Upstream repository https://github.com/flutter/flutter.git
    • Framework revision f1875d570e (7 months ago), 2022-07-13 11:24:16 -0700
    • Engine revision e85ea0e79c
    • Dart version 2.17.6
    • DevTools version 2.12.2

[✓] Android toolchain - develop for Android devices (Android SDK version 33.0.0)
    • Android SDK at ~/Library/Android/sdk
    • Platform android-33, build-tools 33.0.0
    • ANDROID_HOME = ~/Library/Android/sdk
    • ANDROID_SDK_ROOT = ~/Library/Android/sdk
    • Java binary at: /Applications/Android Studio.app/Contents/jre/Contents/Home/bin/java
    • Java version OpenJDK Runtime Environment (build 11.0.12+0-b1504.28-7817840)
    • All Android licenses accepted.

[✓] Xcode - develop for iOS and macOS (Xcode 14.0)
    • Xcode at /Applications/Xcode_14.app/Contents/Developer
    • CocoaPods version 1.11.3

[✓] Chrome - develop for the web
    • Chrome at /Applications/Google Chrome.app/Contents/MacOS/Google Chrome

[✓] Android Studio (version 2021.2)
    • Android Studio at /Applications/Android Studio.app/Contents
    • Flutter plugin can be installed from:
      🔨 https://plugins.jetbrains.com/plugin/9212-flutter
    • Dart plugin can be installed from:
      🔨 https://plugins.jetbrains.com/plugin/6351-dart
    • Java version OpenJDK Runtime Environment (build 11.0.12+0-b1504.28-7817840)

[✓] VS Code (version 1.72.2)
    • VS Code at /Applications/Visual Studio Code.app/Contents
    • Flutter extension version 3.42.0

[✓] Connected device (3 available)
    • iPhone 11 (mobile) • 048D2DE1-4789-4F23-BA2F-A0EE3C1BD78C • ios            • com.apple.CoreSimulator.SimRuntime.iOS-16-0 (simulator)
    • macOS (desktop)    • macos                                • darwin-x64     • macOS 12.6 21G115 darwin-x64
    • Chrome (web)       • chrome                               • web-javascript • Google Chrome 109.0.5414.119

[✓] HTTP Host Availability
    • All required HTTP hosts are available

• No issues found!
```

CHANGELOG.md
## [1.0.0] - 02/01/2023

* First release version.


lib/src/helper.dart
import 'dart:math';

import 'package:flutter/material.dart';

import 'index.dart';

const _kLineBreakMark = '֍';
const _kLineBreakChar = '\n';
const _kSpaceChar = ' ';

/// These modes control how the [AnimatedBouncingText] animates.
enum BouncingTextModes {
  /// Bouncing each character sequentially then stop.
  sequenceOneTime,

  /// Bouncing each character randomly then loop.
  randomlyLoop,

  /// Bouncing each character randomly then stop.
  randomlyOneTime,
}

/// Basically, [AnimatedBouncingText] is a [RichText].
/// So this a helper method is used to create a list of [WidgetSpan].
/// Each item is a [AnimatedBouncingWidget].
/// [mode] : determine how does the animation act.
/// [text] : each character will be convert to [AnimatedBouncingWidget].
/// [textStyle] : see [Text.style].
/// [characterDuration] : duration to animate a character, it will be used as [AnimatedBouncingWidget.duration].
/// [characterDelay] : duration to delay before animating a character.
/// [onEnd] : see [AnimatedBouncingWidget.onEnd].
List<InlineSpan> createAnimatedBouncingSpans({
  required BouncingTextModes mode,
  required String text,
  required TextStyle textStyle,
  required Duration characterDuration,
  required Duration characterDelay,
  Function? onEnd,
}) {
  InlineSpan _buildAWord(
      String word, int paragraphIndex, int wordIndex, int wordsLength) {
    int i = 1;
    return WidgetSpan(
      child: Row(
        mainAxisSize: MainAxisSize.min,
        children: word.characters.map(
          (e) {
            final delayDuration = mode == BouncingTextModes.randomlyLoop ||
                    mode == BouncingTextModes.randomlyOneTime
                ? Duration(
                    milliseconds: Random.secure().nextInt(characterDuration
                            .inMilliseconds +
                        characterDelay.inMilliseconds * (paragraphIndex + i)))
                : Duration(
                    milliseconds:
                        characterDelay.inMilliseconds * (paragraphIndex + i));
            i++;
            final _onCompleted =
                i > word.length && wordIndex == wordsLength - 1 ? onEnd : null;
            return FutureBuilder(
              builder: (context, snapshot) {
                if (snapshot.data != 1) {
                  return SizedBox();
                } else {
                  return AnimatedBouncingWidget(
                    duration: characterDuration,
                    isRepeat: mode == BouncingTextModes.randomlyLoop,
                    child: Text(
                      e,
                      style: textStyle,
                    ),
                    onEnd: _onCompleted,
                  );
                }
              },
              initialData: 0,
              future: Future.delayed(delayDuration, () => 1),
            );
          },
        ).toList(growable: false),
      ),
    );
  }

  InlineSpan _buildALineBreak() => TextSpan(
        text: _kLineBreakChar,
        style: textStyle,
      );

  InlineSpan _buildSpace() => TextSpan(
        text: _kSpaceChar,
        style: textStyle,
      );

  final paragraph = text.replaceAll(
      _kLineBreakChar, '$_kSpaceChar$_kLineBreakMark$_kSpaceChar');
  final words = paragraph.split(_kSpaceChar);
  // final wordsWithoutLineBreak = words.where((e) => e != _kLineBreakMark);
  var spans = <InlineSpan>[];
  int paragraphIndex = 0;
  for (var i = 0; i < words.length; ++i) {
    var word = words[i];
    if (word == _kLineBreakMark) {
      spans.add(_buildALineBreak());
    } else {
      spans.add(_buildAWord(word, paragraphIndex, i, words.length));
      paragraphIndex += word.length;
      if (i < words.length - 1) spans.add(_buildSpace());
    }
  }

  return spans.toList(growable: false);
}


lib/src/animated_bouncing_text.dart
import 'package:flutter/material.dart';

import 'index.dart';

/// A Romantic Bouncing Text
/// Ideally it is a [RichText] to control the paragraph easier.
/// The helper method [createAnimatedBouncingSpans] is used to create its child spans.
/// [mode] : determine how does the animation act.
/// [text] : each character will be convert to [AnimatedBouncingWidget].
/// [textStyle] : see [Text.style].
/// [characterDuration] : duration to animate a character, it will be used as [AnimatedBouncingWidget.duration].
/// [characterDelay] : duration to delay before animating a character.
/// [onEnd] : see [AnimatedBouncingWidget.onEnd].
/// [textAlign] : see [RichText.textAlign].
class AnimatedBouncingText extends StatelessWidget {
  final BouncingTextModes mode;
  final String text;
  final TextStyle textStyle;
  final Duration characterDuration;
  final Duration characterDelay;
  final Function? onEnd;
  final TextAlign textAlign;

  /// Constructor.
  const AnimatedBouncingText({
    Key? key,
    required this.text,
    required this.textStyle,
    this.characterDuration = const Duration(milliseconds: 1000),
    this.characterDelay = const Duration(milliseconds: 300),
    this.mode = BouncingTextModes.sequenceOneTime,
    this.onEnd,
    this.textAlign = TextAlign.start,
  }) : super(key: key);

  @override
  Widget build(BuildContext context) {
    return RichText(
      textAlign: textAlign,
      text: TextSpan(
        style: textStyle,
        children: createAnimatedBouncingSpans(
          text: text,
          textStyle: textStyle,
          mode: mode,
          characterDuration: characterDuration,
          characterDelay: characterDelay,
          onEnd: onEnd,
        ),
      ),
    );
  }
}


lib/src/animated_bouncing_widget.dart
import 'package:flutter/material.dart';

final _kFadeInTween = TweenSequence<double>([
  TweenSequenceItem(tween: Tween<double>(begin: 0, end: 1), weight: 0.1),
  TweenSequenceItem(tween: Tween<double>(begin: 1, end: 1), weight: 0.9),
]);
final _kFallingTween = TweenSequence<double>([
  TweenSequenceItem(tween: Tween<double>(begin: -20, end: 0), weight: 0.1),
  TweenSequenceItem(tween: Tween<double>(begin: 0, end: 0), weight: 0.9),
]);
final _kBouncingTweenX = TweenSequence<double>([
  TweenSequenceItem(tween: Tween<double>(begin: 1, end: 1), weight: 0.1),
  TweenSequenceItem(tween: Tween<double>(begin: 1, end: 1.5), weight: 0.1),
  TweenSequenceItem(tween: Tween<double>(begin: 1.5, end: 0.5), weight: 0.2),
  TweenSequenceItem(tween: Tween<double>(begin: 0.5, end: 1), weight: 0.6),
]);
final _kBouncingTweenY = TweenSequence<double>([
  TweenSequenceItem(tween: Tween<double>(begin: 1, end: 1), weight: 0.1),
  TweenSequenceItem(tween: Tween<double>(begin: 1, end: 0.5), weight: 0.1),
  TweenSequenceItem(tween: Tween<double>(begin: 0.5, end: 1.5), weight: 0.2),
  TweenSequenceItem(tween: Tween<double>(begin: 1.5, end: 1), weight: 0.6),
]);

/// This is an implicitly animated widget. Its [child] will fall down then bounce in [duration].
/// If [isRepeat] is true, the animation will be repeat after completed, else [onEnd] will be called.
class AnimatedBouncingWidget extends StatefulWidget {
  final Widget child;
  final Duration duration;
  final bool isRepeat;
  final Function? onEnd;

  /// Constructor.
  const AnimatedBouncingWidget({
    required this.child,
    Key? key,
    required this.duration,
    this.isRepeat = false,
    this.onEnd,
  }) : super(key: key);

  @override
  State<AnimatedBouncingWidget> createState() => _AnimatedBouncingWidgetState();
}

class _AnimatedBouncingWidgetState extends State<AnimatedBouncingWidget>
    with SingleTickerProviderStateMixin {
  late AnimationController _controller;

  @override
  void didUpdateWidget(covariant AnimatedBouncingWidget oldWidget) {
    if (widget.isRepeat) _start();
    super.didUpdateWidget(oldWidget);
  }

  @override
  void initState() {
    _controller =
        AnimationController(vsync: this, duration: widget.duration, value: 1);
    _controller.addListener(() async {
      if (!mounted) return;
      setState(() {});
      if (widget.isRepeat &&
          _controller.isCompleted &&
          !_controller.isDismissed) {
        _start();
      }
    });
    _controller.addStatusListener((status) {
      if (status == AnimationStatus.completed) {
        if (mounted) {
          widget.onEnd?.call();
        }
      }
    });
    _start();
    super.initState();
  }

  @override
  void dispose() {
    _controller.dispose();
    super.dispose();
  }

  void _start() {
    if (widget.isRepeat) {
      _controller.repeat();
    } else {
      _controller.forward(from: 0);
    }
  }

  @override
  Widget build(BuildContext context) {
    final bouncingValue =
        CurveTween(curve: Curves.bounceOut).transform(_controller.value);
    final dy = _kFallingTween.transform(bouncingValue);
    final sx = _kBouncingTweenX.transform(bouncingValue);
    final sy = _kBouncingTweenY.transform(bouncingValue);
    final fadeInValue =
        CurveTween(curve: Curves.linear).transform(_controller.value);
    final opacity = _kFadeInTween.transform(fadeInValue);
    return Transform.translate(
      offset: Offset(0, dy),
      child: Transform.scale(
        child: Opacity(
          opacity: opacity,
          child: widget.child,
        ),
        scaleX: sx,
        scaleY: sy,
        alignment: Alignment.center,
      ),
    );
  }
}


example/lib/main.dart
import 'package:flutter/material.dart';
import 'package:flutter_bouncing_text/flutter_bouncing_text.dart';

const _kColors = [
  Color(0xff4D4DFF),
  Color(0xff2c85b7),
  Color(0xff693480),
];
void main() => runApp(MyApp());

class MyApp extends StatefulWidget {
  @override
  _MyAppState createState() => _MyAppState();
}

class _MyAppState extends State<MyApp> {
  @override
  Widget build(BuildContext context) {
    return MaterialApp(
      title: 'Example',
      theme: ThemeData(
        primarySwatch: Colors.blue,
      ),
      home: Home(),
    );
  }
}

class Home extends StatefulWidget {
  const Home({Key? key}) : super(key: key);

  @override
  State<Home> createState() => _HomeState();
}

class _HomeState extends State<Home> {
  var characterDuration = Duration(seconds: 1);
  var characterDelay = Duration(seconds: 1);
  var textAlign = TextAlign.start;
  @override
  Widget build(BuildContext context) {
    return Scaffold(
      appBar: AppBar(
        title: Text('Romantic Developer'),
      ),
      body: SafeArea(
        child: SingleChildScrollView(
          padding: const EdgeInsets.all(20.0),
          child: Column(
            crossAxisAlignment: CrossAxisAlignment.stretch,
            children: [
              Padding(
                padding: const EdgeInsets.all(16),
                child: Column(
                  children: [
                    Text(
                      'Text Align',
                      style: TextStyle(
                        fontSize: 14,
                        fontWeight: FontWeight.bold,
                      ),
                    ),
                    Wrap(
                      children: [
                        ...TextAlign.values.map(
                          (e) => Row(
                            children: [
                              Radio<TextAlign>(
                                value: e,
                                groupValue: textAlign,
                                onChanged: (TextAlign? value) {
                                  if (value != null) {
                                    setState(() {
                                      textAlign = value;
                                    });
                                  }
                                },
                              ),
                              Text(
                                e.name,
                                style: TextStyle(
                                  fontSize: 14,
                                ),
                              ),
                            ],
                            mainAxisSize: MainAxisSize.min,
                            crossAxisAlignment: CrossAxisAlignment.center,
                          ),
                        ),
                      ],
                    ),
                    const SizedBox(
                      height: 8,
                    ),
                    Text(
                      'Character Duration',
                      style: TextStyle(
                        fontSize: 14,
                        fontWeight: FontWeight.bold,
                      ),
                    ),
                    Slider(
                      value: characterDuration.inMilliseconds / 1000.0,
                      min: 1,
                      max: 3,
                      divisions: 10,
                      onChanged: (e) => setState(() {
                        characterDuration =
                            Duration(milliseconds: (e * 1000).toInt());
                      }),
                    ),
                    const SizedBox(
                      height: 8,
                    ),
                    Text(
                      'Character Delay',
                      style: TextStyle(
                        fontSize: 14,
                        fontWeight: FontWeight.bold,
                      ),
                    ),
                    Slider(
                      value: characterDelay.inMilliseconds.toDouble(),
                      min: 100,
                      max: 3000,
                      divisions: 20,
                      onChanged: (e) => setState(() {
                        characterDelay = Duration(milliseconds: e.toInt());
                      }),
                    ),
                  ],
                ),
              ),
              ...BouncingTextModes.values.map(
                (e) {
                  final index = BouncingTextModes.values.indexOf(e);
                  return Column(
                    children: [
                      Text(
                        e.name,
                        style: TextStyle(
                          fontSize: 14,
                          fontWeight: FontWeight.bold,
                        ),
                      ),
                      Container(
                        color: _kColors[index],
                        padding: const EdgeInsets.all(20.0),
                        child: AnimatedBouncingText(
                          key: ValueKey(DateTime.now()),
                          text: '($index) I am\nRomantic Developer.',
                          mode: e,
                          textAlign: textAlign,
                          characterDuration: characterDuration,
                          characterDelay: characterDelay,
                          textStyle: TextStyle(
                            color: Colors.white,
                            fontWeight: FontWeight.bold,
                            fontSize: 22,
                          ),
                          onEnd: () {
                            print('_HomeState.build.onEnd.$index');
                          },
                        ),
                      ),
                      const SizedBox(
                        height: 8,
                      ),
                    ],
                  );
                },
              ),
            ],
          ),
        ),
      ),
    );
  }
}
